src/ZkFold/Base/Data/Vector.hs

```diff
@@ -5,110 +5,104 @@
 module ZkFold.Base.Data.Vector where
 
 import           Control.DeepSeq                  (NFData)
-import qualified Control.Monad                    as M
-import           Control.Parallel.Strategies      (parMap, rpar)
+import           Control.Monad.State.Strict       (runState, state)
 import           Data.Aeson                       (ToJSON (..))
-import           Data.Bifunctor                   (first)
 import           Data.Distributive                (Distributive (..))
-import           Data.Functor.Rep                 (Representable (..), collectRep, distributeRep)
-import qualified Data.List                        as List
-import           Data.List.Split                  (chunksOf)
+import           Data.Functor.Rep                 (Representable (..), collectRep, distributeRep, mzipRep, pureRep)
 import           Data.These                       (These (..))
+import qualified Data.Vector                      as V
+import           Data.Vector.Binary               ()
+import qualified Data.Vector.Split                as V
 import           Data.Zip                         (Semialign (..), Zip (..))
 import           GHC.Generics                     (Generic)
-import           Prelude                          hiding (drop, head, length, mod, replicate, sum, tail, take, zip,
-                                                   zipWith, (*))
-import qualified Prelude                          as P
+import           Prelude                          hiding (concat, drop, head, length, mod, replicate, sum, tail, take,
+                                                   zip, zipWith, (*))
 import           System.Random                    (Random (..))
 import           Test.QuickCheck                  (Arbitrary (..))
 
 import           ZkFold.Base.Algebra.Basic.Class
 import           ZkFold.Base.Algebra.Basic.Field
 import           ZkFold.Base.Algebra.Basic.Number
 import           ZkFold.Base.Data.ByteString      (Binary (..))
-import qualified ZkFold.Prelude                   as ZP
-import           ZkFold.Prelude                   (length, replicate)
+import           ZkFold.Prelude                   (length)
 
-newtype Vector (size :: Natural) a = Vector [a]
+newtype Vector (size :: Natural) a = Vector {toV :: V.Vector a}
     deriving (Show, Eq, Functor, Foldable, Traversable, Generic, NFData, Ord)
 
+-- helper
+knownNat :: forall size n . (KnownNat size, Integral n) => n
+knownNat = knownNat @size
```

Comment:
I feel like this is going to loop

Reply:
Yup, I think it suffered a find & replace mishap. Fixed.